File: lib/core/utils/workspace_status.dart
// Copyright (c) 2026 Osman Alperen Çinar-Koraş (oakisnotree). Licensed under AGPL-3.0.

/// Derived workspace status based on its latest run.
///
/// This is a UI-only concept computed from the workspace run's status string.
/// Not part of the Rust SDK domain model.
enum WorkspaceStatus {
  idle,
  starting,
  running,
  stopping,
  failed;

  bool get isTerminal => this == failed;
  bool get isActive => this == starting || this == running || this == stopping;

  /// Parses a run status string into a [WorkspaceStatus].
  /// Returns [idle] if the string doesn't match any known status.
  static WorkspaceStatus fromRunStatus(String? status) {
    return switch (status) {
      'starting' => WorkspaceStatus.starting,
      'running' => WorkspaceStatus.running,
      'stopping' => WorkspaceStatus.stopping,
      'failed' => WorkspaceStatus.failed,
      _ => WorkspaceStatus.idle,
    };
  }
}

File: packages/dspatch_engine/src/domain/enums/workspace_status.rs
// Copyright (c) 2026 Osman Alperen Çinar-Koraş (oakisnotree). Licensed under AGPL-3.0.

use serde::{Deserialize, Serialize};

/// Lifecycle state of a workspace.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WorkspaceStatus {
    Idle,
    Starting,
    Running,
    Stopping,
    Failed,
}

impl WorkspaceStatus {
    /// Whether this is a final state.
    pub fn is_terminal(self) -> bool {
        self == Self::Failed
    }

    /// Whether the workspace is actively doing something.
    pub fn is_active(self) -> bool {
        matches!(self, Self::Starting | Self::Running | Self::Stopping)
    }

    /// Returns `true` if transitioning from this state to `next` is valid.
    pub fn can_transition_to(self, next: WorkspaceStatus) -> bool {
        use WorkspaceStatus::*;
        let valid: &[WorkspaceStatus] = match self {
            Idle => &[Starting],
            Starting => &[Running, Failed],
            Running => &[Stopping, Failed],
            Stopping => &[Idle, Failed],
            Failed => &[Idle],
        };
        valid.contains(&next)
    }
}
